### PR TITLE
feat: use live state for application version

### PR DESCRIPTION
**Pull request by andrii-codefresh**
Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

File: event_reporter/reporter/application_event_reporter.go [hunk 1]
```diff
@@ -143,6 +160,12 @@ func (s *applicationEventReporter) StreamApplicationEvents(
 		return err
```
Comment (dmelnyk-octo) [inline event_reporter/reporter/application_event_reporter.go]: this seems not reachable, can we clean up it?

File: event_reporter/reporter/application_event_reporter.go [hunk 2]
```diff
@@ -143,6 +160,12 @@ func (s *applicationEventReporter) StreamApplicationEvents(
...
 	}
 
+	var applicationVersions *apiclient.ApplicationVersions = nil
+	liveManifests, err, _ := s.getLiveManifests(ctx, a, logCtx)
```
Comment (dmelnyk-octo) [inline event_reporter/reporter/application_event_reporter.go]: can we remove third returned value from function?
it looks like same can be done with `getDesiredState`

File: event_reporter/reporter/application_event_reporter.go [hunk 3]
```diff
@@ -143,6 +160,12 @@ func (s *applicationEventReporter) StreamApplicationEvents(
...
+	if (err == nil) && (liveManifests != nil) && (liveManifests.ApplicationVersions != nil) {
```
Comment (dmelnyk-octo) [inline event_reporter/reporter/application_event_reporter.go]: (liveManifests != nil) always true

Comment (dmelnyk-octo) [reply event_reporter/reporter/application_event_reporter.go]: condition can be simplified to `liveManifests.ApplicationVersions != nil`

**Review by dmelnyk-octo — APPROVED** [review]

File: event_reporter/reporter/application_event_reporter.go [hunk 4]
```diff
@@ -87,13 +87,12 @@ func (s *applicationEventReporter) shouldSendResourceEvent(a *appv1.Application,
...
 }
 
-func (r *applicationEventReporter) getDesiredManifests(ctx context.Context, a *appv1.Application, logCtx *log.Entry) (*apiclient.ManifestResponse, error, bool) {
+func (r *applicationEventReporter) getDesiredManifests(ctx context.Context, a *appv1.Application, logCtx *log.Entry) (*apiclient.ManifestResponse, bool) {
 	// get the desired state manifests of the application
 	project := a.Spec.GetProject()
 	desiredManifests, err := r.applicationServiceClient.GetManifests(ctx, &application.ApplicationManifestQuery{
 		Name:         &a.Name,
 		AppNamespace: &a.Namespace,
-		Revision:     &a.Status.Sync.Revision,
```
Comment (pasha-codefresh) [inline event_reporter/reporter/application_event_reporter.go:96]: why we removed revision from request?

**Review by pasha-codefresh — CHANGES_REQUESTED** [review]

File: event_reporter/reporter/application_event_reporter.go [hunk 5]
```diff
@@ -102,9 +101,26 @@ func (r *applicationEventReporter) getDesiredManifests(ctx context.Context, a *a
...
 		logCtx.WithError(err).Warn("failed to get application desired state manifests, reporting actual state only")
 		desiredManifests = &apiclient.ManifestResponse{Manifests: []*apiclient.Manifest{}}
-		return desiredManifests, nil, true // will ignore requiresPruning=true to not delete resources with actual state
+		return desiredManifests, true // will ignore requiresPruning=true to not delete resources with actual state
 	}
-	return desiredManifests, nil, false
+	return desiredManifests, false
+}
+
+func (r *applicationEventReporter) getLiveManifests(ctx context.Context, a *appv1.Application, logCtx *log.Entry) *apiclient.ManifestResponse {
```
Comment (pasha-codefresh) [inline event_reporter/reporter/application_event_reporter.go]: I can see that here only difference is revision, maybe we should compbine it into single common function? and just pass different input

File: event_reporter/reporter/application_event_reporter.go [hunk 6]
```diff
@@ -132,15 +148,18 @@ func (s *applicationEventReporter) StreamApplicationEvents(
...
 	logCtx.Info("getting desired manifests")
 
-	desiredManifests, err, manifestGenErr := s.getDesiredManifests(ctx, a, logCtx)
-	if err != nil {
-		return err
+	desiredManifests, manifestGenErr := s.getDesiredManifests(ctx, a, logCtx)
+
+	var applicationVersions *apiclient.ApplicationVersions = nil
+	liveManifests := s.getLiveManifests(ctx, a, logCtx)
+	if liveManifests.ApplicationVersions != nil {
```
Comment (pasha-codefresh) [inline event_reporter/reporter/application_event_reporter.go]: you can just do liveManifests.GetApplicationVersions() and you dont need this if, also you dont need create variable in such case

File: pkg/version_config_manager/version_config_manager.go [hunk 7]
```diff
@@ -13,6 +13,9 @@ type VersionConfig struct {
 	ResourceName string `json:"resourceName"`
 }
 
+var DefaultVersionSource = "Chart.yaml"
```
Comment (pasha-codefresh) [inline pkg/version_config_manager/version_config_manager.go]: please change to const , like this

const (
  variableName = ....
)

File: reposerver/repository/app_version.go [hunk 8]
```diff
@@ -61,8 +61,16 @@ func getVersionFromFile(appPath, jsonPathExpression string) (*string, error) {
 	if err != nil {
 		return nil, err
 	}
-	appVersion, ok := versionValue.(string)
-	if !ok {
+
+	var appVersion string
+	var conversionSuccess bool
+	if versionArray, ok := versionValue.([]interface{}); ok && len(versionArray) > 0 {
```
Comment (pasha-codefresh) [inline reposerver/repository/app_version.go]: could you please move it to dedicated function that will return two values and cover it with primitive unit test. It is pretty sensitive code that may cause crash of app if something will went wrong

**Review by pasha-codefresh — CHANGES_REQUESTED** [review]

File: event_reporter/reporter/application_event_reporter.go (hunk 4 above)
Comment (andrii-codefresh) [reply event_reporter/reporter/application_event_reporter.go:96]: The request parameters for retrieving the manifest have changed. Previously, the revision was not passed in the request, and the manifest was always generated from the latest commit of the branch. Now the revision is passed, so we need to remove the revision from the function call to maintain the same behavior.

**Review by dmelnyk-octo — APPROVED** [review]

**Pull request merged**